src/sql/SqlResult.ts

```diff
@@ -0,0 +1,380 @@
+        if (this.returnRawResult) { // Return SqlRow
+            const columnCount = this.currentPage.getColumnCount();
+            const values = new Array(columnCount);
+            for (let i = 0; i < columnCount; i++) {
+                values[i] = this.serializationService.toObject(this.currentPage.getValue(this.currentPosition, i));
+            }
+            return new SqlRowImpl(values, this.rowMetadata);
+        } else { // Return objects
+            const result: SqlRowAsObject = {};
+            for (let i = 0; i < this.currentPage.getColumnCount(); i++) {
+                const columnMetadata = this.rowMetadata.getColumnByIndex(i);
+                result[columnMetadata.name] = this.serializationService.toObject(
+                    this.currentPage.getValue(this.currentPosition, i)
+                );
+            }
+            return result;
+        }
+    }
+
+    /** Called when a execute response is received. */
+    onExecuteResponse(rowMetadata: SqlRowMetadata | null, rowPage: SqlPage, updateCount: Long) {
+        // Ignore the response if SQL result is closed.
+        if (this.closed) return;
+
+        if (rowMetadata !== null) { // Result that including rows
+            this.rowMetadata = rowMetadata;
+            this.onNextPage(rowPage);
+            this.updateCount = Long.fromInt(-1);
+        } else { // Result that including update count
+            this.updateCount = updateCount;
+            this.closed = true;
+        }
+        this.executeDeferred.resolve(true);
+    }
+
+    /**
+     * Fetches the next page. Called internally by iteration logic. Pages are fetched whenever the current page is fully
+     * iterated.
+     */
+    fetch(): Promise<SqlPage> {
+        // If there is an ongoing fetch, return that promise
+        if (this.fetchDeferred?.promise)
+            return this.fetchDeferred.promise;
+
+        // Do not start a fetch if close is called
+        if (this.closeDeferred?.promise) {
+            return Promise.reject('Cannot fetch, close() is called');
+        }
+        // Do not start a fetch if the result is already closed
+        if (this.closed) {
+            return Promise.reject('Cannot fetch, the result is already closed');
+        }
+
+        this.fetchDeferred = deferredPromise<SqlPage>();
+
+        this.sqlService.fetch(this.connection, this.queryId, this.cursorBufferSize).then(value => {
+            this.fetchDeferred.resolve(value);
+            this.fetchDeferred = undefined; // Set fetchDeferred to undefined to be able to fetch again
+        }).catch(err => {
+            this.fetchDeferred.reject(this.sqlService.toHazelcastSqlException(err, this.connection));
+        });
+
+        return this.fetchDeferred.promise;
+    }
+
+    /** Returns if there are rows to be iterated. Used by {@link next}. */
+    hasNext(): Promise<boolean> {
+        return this.executeDeferred.promise.then(() => {
+            if (this.updateCount !== Long.fromInt(-1)) {
```

Comment:
Since `hasNext` is called a lot, I think we can either
- make the `Long.fromInt(-1)` a constant and check against it, instead of recreating it each time
- use `rowMetadata` with null check instead of `updateCount`

I think second option should be faster